e2e/command_test.go

```diff
@@ -0,0 +1,233 @@
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul-terraform-sync/api"
+	"github.com/hashicorp/consul-terraform-sync/testutils"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestE2E_MetaCOmmandErrors(t *testing.T) {
+	// test cases that cross subcommands that coded in the command meta object
+	t.Parallel()
+
+	srv, err := newTestConsulServer(t)
+	require.NoError(t, err, "failed to start consul server")
+	defer srv.Stop()
+
+	tempDir := fmt.Sprintf("%s%s", tempDirPrefix, "meta_errs")
+	delete, err := testutils.MakeTempDir(tempDir)
+	// no defer to delete directory: only delete at end of test if no errors
+	require.NoError(t, err)
+
+	configPath := filepath.Join(tempDir, configFile)
+
+	port, err := api.FreePort()
+	require.NoError(t, err)
+
+	err = makeConfig(configPath, fakeHandlerConfig(srv.HTTPAddr, tempDir, port))
+	require.NoError(t, err)
+
+	cmd, err := runSyncDevMode(configPath)
+	defer stopCommand(cmd)
+	require.NoError(t, err)
+	time.Sleep(5 * time.Second)
+
+	cases := []struct {
+		name           string
+		args           []string
+		outputContains string
+	}{
+		{
+			"missing required arguments",
+			[]string{},
+			"Error: this command requires one argument",
+		},
+		{
+			"unsupported argument",
+			[]string{"-unsupported-flag", fakeFailureTaskName},
+			"Error: unsupported arguments in flags",
+		},
+		{
+			"non-existing task",
+			[]string{fmt.Sprintf("-port=%d", port), "non-existent-task"},
+			"does not exist or has not been initialized yet",
+		},
+		{
+			"out of order arguments",
+			[]string{fakeFailureTaskName, fmt.Sprintf("-port %d", port)},
+			"All flags are required to appear before positional arguments",
+		},
+	}
+
+	for _, lifecycle := range []string{"disable", "enable"} {
+		// run through all the test cases for each task lifcycle command
+		for _, tc := range cases {
+			t.Run(fmt.Sprintf("%s/%s", lifecycle, tc.name), func(t *testing.T) {
+				subcmd := []string{"task", lifecycle}
+				subcmd = append(subcmd, tc.args...)
+
+				output, err := runSubcommand(t, "", subcmd...)
+				assert.Contains(t, output, tc.outputContains)
+				assert.Error(t, err)
+			})
+		}
+	}
+
+	delete()
+}
+
+func TestE2E_EnableTaskCommand(t *testing.T) {
+	t.Parallel()
+
+	srv, err := newTestConsulServer(t)
+	require.NoError(t, err, "failed to start consul server")
+	defer srv.Stop()
+
+	tempDir := fmt.Sprintf("%s%s", tempDirPrefix, "enable_cmd")
+	delete, err := testutils.MakeTempDir(tempDir)
+	// no defer to delete directory: only delete at end of test if no errors
+	require.NoError(t, err)
+
+	configPath := filepath.Join(tempDir, configFile)
+
+	port, err := api.FreePort()
+	require.NoError(t, err)
+
+	err = makeConfig(configPath, disabledTaskConfig(srv.HTTPAddr, tempDir, port))
+	require.NoError(t, err)
+
+	cmd, err := runSyncDevMode(configPath)
+	defer stopCommand(cmd)
+	require.NoError(t, err)
+	time.Sleep(5 * time.Second)
```

Comment:
For API-dependent e2e tests, I can see a future improvement to have the CTS client support a health endpoint to verify that the CLI is good to interact with API server.

Reply:
Great idea! I'd love that compared to these arbitrary sleeps

Reply:
It'll still poll/sleep probably but yeah a bit less arbitrary 🙂